jobs/base_model_validator.py

```diff
@@ -0,0 +1,161 @@
+
+        Args:
+            regex_string: str. A regex pattern for valid ids.
+        """
+        super(ValidateModelIdWithRegex, self).__init__()
+        self.regex = re.compile(regex_string)
+
+    def process(self, input_model):
+        """Function that defines how to process each element in a pipeline of
+        models.
+
+        Args:
+            input_model: datastore_services.Model. Entity to validate.
+
+        Yields:
+            ModelInvalidIdError. An error class for models with invalid IDs.
+        """
+        model = jobs_utils.clone_model(input_model)
+
+        if not self.regex.match(model.id):
+            yield errors.ModelInvalidIdError(model)
+
+
+class ValidateDeleted(beam.DoFn):
+    """DoFn to check whether models marked for deletion are stale."""
+
+    def process(self, input_model):
+        """Function that defines how to process each element in a pipeline of
+        models.
+
+        Args:
+            input_model: datastore_services.Model. Entity to validate.
+
+        Yields:
+            ModelExpiredError. An error class for expired models.
+        """
+        model = jobs_utils.clone_model(input_model)
+        date_now = datetime.datetime.utcnow()
+
+        expiration_date = (
+            date_now -
+            cron_services.PERIOD_TO_HARD_DELETE_MODELS_MARKED_AS_DELETED)
+
+        if model.last_updated < expiration_date:
+            yield errors.ModelExpiredError(model)
+
+
+class ValidateModelTimeFields(beam.DoFn):
+    """DoFn to check whether created_on and last_updated timestamps are
+    valid."""
+
+    def process(self, input_model):
+        """Function that defines how to process each element in a pipeline of
+        models.
+
+        Args:
+            input_model: datastore_services.Model. Entity to validate.
+
+        Yields:
+            ModelMutatedDuringJobError. Error for timestamp validation.
+            ModelTimestampRelationshipError. Error for timestamp validation.
+        """
+        model = jobs_utils.clone_model(input_model)
+        if model.created_on > (model.last_updated + MAX_CLOCK_SKEW_SECS):
+            yield errors.ModelTimestampRelationshipError(model)
+
+        current_datetime = datetime.datetime.utcnow()
+        if (model.last_updated - MAX_CLOCK_SKEW_SECS) > current_datetime:
+            yield errors.ModelMutatedDuringJobError(model)
+
+
+class BaseModelValidator(beam.PTransform):
+    """Composite beam Transform which returns a pipeline of validation
+    errors."""
+
+    def expand(self, model_pipe):
+        """Function that takes in a beam.PCollection of datastore models and
+        returns a beam.PCollection of validation errors.
+
+        Args:
+            model_pipe: beam.PCollection. A collection of models.
+
+        Returns:
+            beam.PCollection. A collection of errors represented as
+            key-value pairs.
+        """
+        deleted, not_deleted = (
+            model_pipe | 
+            beam.Map(
```

Comment:
Please put pipes at the beginning of a line (here and elsewhere):

```suggestion
            model_pipe
            | beam.Map(
```

Reply:
This is fixed.